drivers/driver.go

```diff
@@ -149,6 +149,19 @@ type LayerIDMapUpdater interface {
 	SupportsShifting() bool
 }
 
+// AdditionalLayerStore is the directory where exploded layer diff is stored.
+// Each layer can be accessed at:
+// Path/base64(RequiredAnnotations[0]=<value>)/.../base64(RequiredAnnotations[len(RequiredAnnotations)]=<value>)/
```

Comment:
why do we need the annotations as part of the path?

What is the disadvantage of using directly the layer digest as specified in the manifest?

Reply:
@giuseppe I initally thought we can use annotations for content verification: passing the digest of TOC from the runtime to the filesystem for ensuring that the (lazily) pulled contents is ones expected by the runtime. But I currently think we can let the filesystem do all these tasks (i.e. fetching manifest, getting TOC digest and verifying each chunk all on the filesystem's side).
Should we limit the possible path elements to `reference` and `layerdigest`?

Additionally, I thought we don't need `key=` in the part of the path because the order of `value` is deterministic (specified through driver's config `layerstore=<ALS root directory>:<key1>:<key2>:...:<keyX>`). So the path can be the following instead:
```
<ALS root>/base64("value1")/base64("value2")/.../base64("valueX")/
`-- diff
`-- info
```

WDYT?

Reply:
> @giuseppe I initally thought we can use annotations for content verification: passing the digest of TOC from the runtime to the filesystem for ensuring that the (lazily) pulled contents is ones expected by the runtime. But I currently think we can let the filesystem do all these tasks (i.e. fetching manifest, getting TOC digest and verifying each chunk all on the filesystem's side).

I think this is a good idea.  I didn't know of why you were doing it.

My only comment is about storing this kind of metadata somewhere else (maybe just a json file?), so that in future we can extend it if we need, and that will simplify the path to just store the digest without any annotation.

```
<ALS root>/checksum-for-the-layer/
-- diff
-- info
-- info/metadata
```

That would work in a similar way to what I am doing with the Layers Big Data PR.

WDYT?

Reply:
@giuseppe Thanks for the suggestion. Please let me clarify my understanding about `info/metadata`. Which of the following did you mean?

A) The runtime writes some additional information (TOC digest, in this case) to `info/metadata` then access to `diff`. The filesystem reads `info/metadata` written by the runtime, does some additional tasks (verifying TOC, in this case) and shows the exploded diff to `diff` directory.

B) The filesystem shows `info/metadata` containing some additional information (TOC digest, in this case) to the runtime. The runtime does some tasks (verifying TOC, in this case) before it uses `diff` directory. (In this case, image reference still needs to be contained in the path for enabling the filesystem to query layer data to the registry)

Reply:
Should the runtime store such information?   Would it not be the file system responsibility to fetch the TOC digest and do the verification?

I was thinking how the file system could expose this metadata to the runtime, if the runtime needs to access it (and we could probably look at storing it in a format compatible with the SetBigData PR I am working on)

Reply:
> Would it not be the file system responsibility to fetch the TOC digest and do the verification?

I agree with you. We should make verification filesystem's responsibility.

> I was thinking how the file system could expose this metadata to the runtime, if the runtime needs to access it (and we could probably look at storing it in a format compatible with the SetBigData PR I am working on)

Currently I don't immediately come up with the usecase where the fileystem needs to expose metadata (because if we  do the verification on the filesystem's side, we don't need to expose TOC digest to the runtime anymore). But if we'll need to do this, I think we can make it compatible to layer BigData in the way you mentioned in https://github.com/containers/storage/pull/795#discussion_r560063580.

Reply:
thanks.

Yes, so I think we can avoid exposing metadata for now.  If it is ever required we can add it later.

Reply:
> We should make verification filesystem's responsibility.

So how is that going to work? The data available to `PutFromAdditionalLayerStore` is a single digest; does the filesystem read all of the layer to verify the digest? If I understand correctly, that would defeat the entire purpose.

Reply:
The file system starts with the TOC and its digest (through an annotation).  Once the TOC is verified, each file is referenced from there and can be verified separately.

@ktock any further details you'd like to add?

Reply:
nevermind my last question, I see you added a comment with more details